### PR TITLE
[Fix] The interface `multiscale_output` is defined but not used 

### PR DESCRIPTION
**Pull request by AronLin**
The interface `multiscale_output` is defined in `_make_stage` but never used.

HRNet should be allowed to output a one-level feature in stage 4.
So we should use `multiscale_output` while making stage 4 and add interface  `multiscale_output` in the initialization function of HRNet.

Add unit test for HRNet in this PR, too.

### CODE REVIEW

**Comment by AronLin** [comment]
@Junjun2016, please have a look, there is the same problem in mmseg.

File: mmdet/models/backbones/hrnet.py [hunk 1]
```diff
@@ -218,6 +218,8 @@ class HRNet(BaseModule):
             memory while slowing down the training speed.
         zero_init_residual (bool): whether to use zero init for last norm layer
             in resblocks to let them behave as identity.
+        multiscale_output (bool): whether to output multi-level features. If
+            False, only the first level feature will be output. Default: True.
```
Comment (Junjun2016) [inline mmdet/models/backbones/hrnet.py:222]: Add unittests for this `arg`.

Comment (Junjun2016) [reply mmdet/models/backbones/hrnet.py:222]: level --> branch?

**Review by Junjun2016 — COMMENTED** [review]
It seems that unittests for HRNet are missing.

File: mmdet/models/backbones/hrnet.py (hunk 1 above)
Comment (Junjun2016) [reply mmdet/models/backbones/hrnet.py:222]: Add more details for extra `args`.

**Comment by Junjun2016** [comment]
Please update the title.

File: mmdet/models/backbones/hrnet.py (hunk 1 above)
Comment (ZwwWayne) [reply mmdet/models/backbones/hrnet.py:222]: > level --> branch?

should use level. Multi-level features produced by multiple branches

File: mmdet/models/backbones/hrnet.py [hunk 2]
```diff
@@ -218,6 +218,8 @@ class HRNet(BaseModule):
             memory while slowing down the training speed.
         zero_init_residual (bool): whether to use zero init for last norm layer
             in resblocks to let them behave as identity.
+        multiscale_output (bool): whether to output multi-level features. If
```
Comment (hhaAndroid) [inline mmdet/models/backbones/hrnet.py:221]: whether -> Whether and all other docstr.

File: mmdet/models/backbones/hrnet.py (hunk 1 above)
Comment (Junjun2016) [reply mmdet/models/backbones/hrnet.py:222]: > > level --> branch?
> 
> should use level. Multi-level features produced by multiple branches

So it's better to explain it.

> Multi-level features produced by multiple branches

File: tests/test_models/test_backbones/test_hrnet.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import pytest
+import torch
+
+from mmdet.models.backbones.hrnet import HRNet
+
+
+def test_hourglass_backbone():
```
Comment (Junjun2016) [inline tests/test_models/test_backbones/test_hrnet.py:7]: def test_hrnet():

File: mmdet/models/backbones/hrnet.py [hunk 4]
```diff
@@ -207,20 +207,31 @@ class HRNet(BaseModule):
...
             in resblocks to let them behave as identity.
-        multiscale_output (bool): whether to output multi-level features. If
-            False, only the first level feature will be output. Default: True.
-        pretrained (str, optional): model pretrained path. Default: None
+        multiscale_output (bool): Whether to output multi-level features
+            produced by multiple branches. If False, only the first level
+            feature will be output. Default: True.
+        pretrained (str, optional): Model pretrained path. Default: None
         init_cfg (dict or list[dict], optional): Initialization config dict.
             Default: None
```
Comment (Junjun2016) [inline mmdet/models/backbones/hrnet.py:237]: Default: None --> Default: None.

File: mmdet/models/backbones/hrnet.py [hunk 5]
```diff
@@ -207,20 +207,31 @@ class HRNet(BaseModule):
     arXiv: https://arxiv.org/abs/1904.04514
```
Comment (Junjun2016) [inline mmdet/models/backbones/hrnet.py:207]: The citation style is not correct.
Please refer to https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/mobilenet_v3.py#L17.

**Comment by Junjun2016** [comment]
Unitests for other `module` in HRNet is also necessary.
Please refer to unitests for ResNet.

**Comment by AronLin** [comment]
kindly ping @Junjun2016.

**Comment by Junjun2016** [comment]
> kindly ping @Junjun2016.

LGTM

File: mmdet/models/backbones/hrnet.py [hunk 6]
```diff
@@ -203,24 +203,37 @@ def forward(self, x):
...
-        norm_cfg (dict): dictionary to construct and config norm layer.
+        conv_cfg (dict): Dictionary to construct and config conv layer.
+        norm_cfg (dict): Dictionary to construct and config norm layer.
         norm_eval (bool): Whether to set norm layers to eval mode, namely,
             freeze running stats (mean and var). Note: Effect on Batch Norm
             and its variants only.
         with_cp (bool): Use checkpoint or not. Using checkpoint will save some
             memory while slowing down the training speed.
-        zero_init_residual (bool): whether to use zero init for last norm layer
+        zero_init_residual (bool): Whether to use zero init for last norm layer
```
Comment (jshilong) [inline mmdet/models/backbones/hrnet.py:229]: Default value

**Review by jshilong — APPROVED** [review]

File: mmdet/models/backbones/hrnet.py [hunk 7]
```diff
@@ -203,24 +203,37 @@ def forward(self, x):
...
+        extra (dict): Detailed configuration for each stage of HRNet.
+            There must be 4 stages, the configuration for each stage must have
+            5 keys:
+
+                - num_modules(int): The number of HRModule in this stage.
+                - num_branches(int): The number of branches in the HRModule.
+                - block(str): The type of convolution block.
+                - num_blocks(tuple): The number of blocks in each branch.
+                    The length must be equal to num_branches.
+                - num_channels(tuple): the number of channels in each branch.
```
Comment (hhaAndroid) [inline mmdet/models/backbones/hrnet.py:219]: the number of

**Review by ZwwWayne — APPROVED** [review]

**Review by Junjun2016 — APPROVED** [review]

**Pull request merged**